primitives/src/merkle_tree.rs

```diff
@@ -0,0 +1,186 @@
+use merkletree::hash::Algorithm;
+use merkletree::merkle;
+use merkletree::merkle::VecStore;
+use merkletree::proof::Proof;
+use std::hash::Hasher;
+use std::iter::FromIterator;
+use tiny_keccak::Keccak;
+
+#[derive(Clone)]
+struct KeccakAlgorithm(Keccak);
+
+impl KeccakAlgorithm {
+    pub fn new() -> KeccakAlgorithm {
+        KeccakAlgorithm(Keccak::new_keccak256())
+    }
+}
+
+impl Default for KeccakAlgorithm {
+    fn default() -> KeccakAlgorithm {
+        KeccakAlgorithm::new()
+    }
+}
+
+impl Hasher for KeccakAlgorithm {
+    #[inline]
+    fn write(&mut self, msg: &[u8]) {
+        self.0.update(msg)
+    }
+
+    #[inline]
+    fn finish(&self) -> u64 {
+        unimplemented!()
+    }
+}
+
+type MerkleItem = [u8; 32];
+
+impl Algorithm<MerkleItem> for KeccakAlgorithm {
+    #[inline]
+    fn hash(&mut self) -> MerkleItem {
+        let mut res: [u8; 32] = [0; 32];
+        self.0.clone().finalize(&mut res);
+        res
+    }
+
+    #[inline]
+    fn reset(&mut self) {
+        self.0 = Keccak::new_keccak256()
+    }
+
+    fn leaf(&mut self, leaf: MerkleItem) -> MerkleItem {
+        leaf
+    }
+
+    fn node(&mut self, left: MerkleItem, right: MerkleItem, _height: usize) -> MerkleItem {
+        let left_vec: Vec<u8> = left.to_vec();
+        let right_vec: Vec<u8> = right.to_vec();
+
+        let mut node_vec = vec![left_vec, right_vec];
+        node_vec.sort();
+
+        let flatten_node_vec: Vec<u8> = node_vec.into_iter().flatten().collect();
+
+        self.write(&flatten_node_vec.as_slice());
```

Comment:
I think you can just do `self.write(&flatten_node_vec);` the Vec implements the same thing as the String, so it should Deref it to a slice